src/main/java/de/tum/in/www1/artemis/service/ProgrammingExerciseTestCaseService.java

```diff
@@ -0,0 +1,158 @@
+            }
+
+            @Override
+            public boolean isAuthenticated() {
+                return false;
+            }
+
+            @Override
+            public void setAuthenticated(boolean isAuthenticated) throws IllegalArgumentException {
+
+            }
+
+            @Override
+            public String getName() {
+                return null;
+            }
+        };
+    }
+
+    /**
+     * Returns all test cases for a programming exercise.
+     * 
+     * @param id of a programming exercise.
+     * @return test cases of a programming exercise.
+     */
+    public Set<ProgrammingExerciseTestCase> findByExerciseId(Long id) {
+        SecurityContext context = SecurityContextHolder.getContext();
+        context.setAuthentication(getAuthDummy());
+        return this.testCaseRepository.findByExerciseId(id);
+    }
+
+    /**
+     * Returns all active test cases for a programming exercise. Only active test cases are evaluated on build runs.
+     * 
+     * @param id of a programming exercise.
+     * @return active test cases of a programming exercise.
+     */
+    public Set<ProgrammingExerciseTestCase> findActiveByExerciseId(Long id) {
+        SecurityContext context = SecurityContextHolder.getContext();
+        context.setAuthentication(getAuthDummy());
+        return this.testCaseRepository.findActiveByExerciseId(id);
+    }
+
+    /**
+     * From a list of build run feedback, extract all test cases. If an already stored test case is not found anymore in the build result, it will not be deleted, but set inactive.
+     * This way old test cases are not lost, some interfaces in the client might need this information to e.g. show warnings. Returns true if the test cases have changed, false if
+     * they haven't.
+     * 
+     * @param feedbacks list of build log output.
+     * @param exercise  programming exercise.
+     */
+    public boolean generateFromFeedbacks(List<Feedback> feedbacks, ProgrammingExercise exercise) {
+        SecurityContext context = SecurityContextHolder.getContext();
+        context.setAuthentication(getAuthDummy());
+
+        Set<ProgrammingExerciseTestCase> existingTestCases = testCaseRepository.findByExerciseId(exercise.getId());
+        Set<ProgrammingExerciseTestCase> testCasesFromFeedbacks = feedbacks.stream()
+                .map(feedback -> new ProgrammingExerciseTestCase().testName(feedback.getText()).weight(1).exercise(exercise).active(true)).collect(Collectors.toSet());
+        // Get test cases that are not already in database - those will be added as new entries.
+        Set<ProgrammingExerciseTestCase> newTestCases = testCasesFromFeedbacks.stream().filter(testCase -> existingTestCases.stream().noneMatch(testCase::equals))
+                .collect(Collectors.toSet());
+        // Get test cases which activate state flag changed.
+        Set<ProgrammingExerciseTestCase> activationStateChanges = existingTestCases.stream().filter(existing -> {
```

Comment:
activationStateChanges --> I would also include the term "TestCases" in the variable

Reply:
True!